### DISCUSSION TITLE
Sending audio to snapserver from remote device?

### DISCUSSION THREAD

**Discussion by typhoon2099** [snapcast/snapcast#1031] [Q&A]
I'm trying to send audio from my laptop (mpd) to my raspberry pi (snapserver). I think I can send the audio to snapserver using tcp, but I'm not sure how to do this from mpd. I've set up the server with a tcp pipe in server mode,  but I can't find a config for sending audio to the tcp address.

Has anyone else set up something similar?

**Comment by mrwsl** [comment]
Maybe this is helpful: https://wiki.archlinux.org/title/Music_Player_Daemon/Tips_and_tricks#Remote

**Comment by typhoon2099** [comment]
Hmm, Pulse audio might be the way to get here. I'll just have to figure out how it pipes into snapserver without outputting directly, but sounds like a plausible solution. I was hoping for something native to snapcast but I guess one more thing between output and input isn't too bad.

UPDATE: I've configured PulseAudio and it's working quite well. Slightly harder than I thought it would be but mostly because I was doing it wrong.

**Reply by jennussbaum** [reply]
Would you be willing to share your PulseAudio configuration, or to describe what you did? I was trying to set something like that up, but gave up. I know I'm doing it wrong too :-(

**Reply by typhoon2099** [reply]
I'm not at my laptop at the moment, but the rough idea is that MPD outputs to PulseAudio and is configured to point at the Raspberry Pi running Snapserver. That Raspberry Pi has PulseAudio installed, and I think I have it set up to output to a named pipe by default. Snapserver is the only thing running on the Pi so I'm not worried about anything else needing to output properly.

That named pipe is then picked up by Snapserver as a source, and distributed normally (including on the Pi through a snapclient on there).

I can provide specific configs later on today if that doesn't make enough sense or you've already understood it at that level but can't get it to communicate properly.

**Reply by jennussbaum** [reply]
I'd be very grateful if you could, in fact, provide specific configs for that. I did understand the general idea, but was unable to get it working in practice. It always seemed as if it would be possible, so I'm very glad to learn that you accomplished it, but a push in the right direction would be most appreciated. Thank you!

**Reply by typhoon2099** [reply]
Okay, so in `/etc/pulse/default.pa` I have added:

```
load-module module-pipe-sink file=/tmp/snapfifo rate=48000
```

In `~/.pulse/default.pa`:

```
.include /etc/pulse/default.pa

set-sink-volume fifo_output 65536
```

`/etc/snapserver.conf`:

```
source = pipe:///tmp/snapfifo?name=Default
```

All of that is on the Raspberry Pi. On my laptop I have MPD configured to use the Pulse Audio output:

```
audio_output {
  type "pulse"
  name "Snapcast (PulseAudio)"
  server "living-room-audio.local"
  sink "fifo_output"
  format "48000:16:2"
}
```

And that should be it. Feel free to compare what you've got and what I've got and if something doesn't make sense then let me know.

As I was setting this up I had a terminal open looking at the named pipe with `tail -f /tmp/snapfifo`. If the output is going crazy then the audio is making it to the server and PulseAudio is set up correctly. If you can't hear audio then it means that snapserver isn't configured correctly.

Hope that helps!
